content/7-webapp-features/1-fetch-data/ember/users.js

```diff
@@ -0,0 +1,17 @@
+import Route from '@ember/routing/route';
+
+export default class UsersRoute extends Route {
+	async model() {
```

Comment:
IsLoading state will never be observed due to how 'model' is async and will not resolve until all sub async functions are resolved themselves.

If you want same-template loading state, you Don't want to use a route file at all. (They allow you to specify a loading template, but that may oc may not be whan we want, since this site is comparing component patterns.

Reply:
Yeah, I think just removing isLoading will be good?

Reply:
Nah, because all the other examples have isLoading.

I think we need to do the whole example in a component

Reply:
We could use the loading route substate, but that's not a component either 🙃

Reply:
Yeah, so what would feasible?

Reply:
The absolute easiest way to fetch data is with a library.

Examples:
 - https://ember-resources.pages.dev/modules/util_remote_data
 - https://github.com/chriskrycho/ember-async-data
 
 
The main problem is that ember doesn't _yet_ have resources as a primitive:
 - https://ember-resources.pages.dev/modules/util_function_resource (these examples build `RemoteData`)
 - Polaris sketches: https://wycats.github.io/polaris-sketchwork/reactivity.html
 
 There is a talk on this subject that was just at the last emberconf: https://www.youtube.com/watch?v=9_juemvnZVo
 
 
 Your options:
 
- initiate a fetch from the constructor, lose the ability to refetch
- initiate a fetch from the template via helper or modifier -- semantically this is super goofy, however, you could get clever with _with a bunch of extra code_.
- use ember-concurrency to manage loading / error / concurrent problems, but still have the prior to problems to solve
- use a library specifically designed for reactive data-fetching.

Since, Svelte and React made their own boilerplate-heavy abstraction for fetching data, I don't think it'd be too bonkers to provide the full thing in ember (today)

but I think it would be good to somehow show that common libraries make things way easier. idk.

Here is the code I was thinking if we just did this pure vanilla (and this very closely matches what Svelte and React are doing in Component Party today):
```ts
import { tracked } from '@glimmer/tracking';
import Component from '@glimmer/component';

export default class Demo extends Component {
  fetchUsers = () => getUsers();
}

class State {
  @tracked isLoading = false;
  @tracked error = null;
  @tracked data = null;
}

function getUsers() {
  let state = new State();
  
  async function fetchData() {
    state.isLoading = true;
    
    try {
      let response = await fetch('https://randomuser.me/api/?results=3');
      let { results: users } = await response.json();
      state.data = users;
      state.error = null;  
    } catch (err) {
      state.data = null;
      state.error = err;
    }
    
    state.isLoading = false;
  }

  fetchData();

  return state;
}
```

```hbs
{{#let (fetchUsers) as |request|}}
  {{#if request.isLoading}}
    <p>Fetching users...</p>
  {{else if request.error}}
    <p>An error occurred while fetching users</p>
  {{else}}
     <ul>
       {{#each request.users as |user|}}
          <li>
            <img src={{user.picture.thumbnail}} alt="user">
            <p>{{user.name.first}} {{user.name.last}}</p>
          </li>
       {{/each}}
     </ul>
  {{/if}}
  
{{/let}}
```

Reply:
Of course it would be simpler with libraries but the purpose is to use framework features to show a "behind the scenes" of libraries. Your proposal is exactly matching Component Party philosophy.
Do you mind creating a PR with your proposal ?